### PR TITLE
Add more functions to `dual.hpp`

### PR DESCRIPTION
**Pull request by jamiebramwell**
This adds asin, acos, atan, min, and max to the dual class.

ping @kswartz92

### CODE REVIEW

File: src/serac/numerics/functional/tests/test_tensor_ad.cpp [hunk 1]
```diff
@@ -20,14 +20,45 @@ TEST(Tensor, Norm)
...
+{
+  auto xd = max(make_dual(x), make_dual(y));
+  EXPECT_DOUBLE_EQ(std::max(x, y), xd.value);
+}
+
+TEST(DualNumberTensor, Min)
+{
+  auto xd = min(make_dual(x), make_dual(y));
+  EXPECT_DOUBLE_EQ(std::min(x, y), xd.value);
+}
```
Comment (samuelpmishLLNL) [inline src/serac/numerics/functional/tests/test_tensor_ad.cpp]: These tests should also check that the derivative has the correct value. i.e.

```cpp
const double x = 0.5;
const double y = 0.25;
...
auto xd = max(dual{x, 1.0}, dual{y, 2.0});
EXPECT_DOUBLE_EQ(x, xd.value);
EXPECT_DOUBLE_EQ(1.0, xd.gradient);
```

File: src/serac/numerics/functional/dual.hpp [hunk 2]
```diff
@@ -206,6 +206,48 @@ SERAC_HOST_DEVICE auto abs(dual<gradient_type> x)
...
+{
+  return (a.value > b) ? a : b;
+}
+
+/** @brief implementation of max for dual numbers */
+template <typename gradient_type>
+SERAC_HOST_DEVICE auto max(double a, dual<gradient_type> b)
+{
+  return (a > b.value) ? a : b;
+}
```
Comment (samuelpmishLLNL) [inline src/serac/numerics/functional/dual.hpp]: We should add tests for these mixed dual/double overloads. I don't think this implementation compiles-- the different branches of the ternary operator have different types.

Same for the dual/double `min` implementation.

**Comment by jamiebramwell** [comment]
I think this is ready to go now. Any other comments @samuelpmishLLNL @white238 ?

**Review by white238 — APPROVED** [review]
Approved, pending random documentation thought

File: src/serac/numerics/functional/dual.hpp [hunk 3]
```diff
@@ -199,13 +199,80 @@ SERAC_HOST_DEVICE constexpr auto& operator-=(dual<gradient_type>& a, double b)
...
+ */
 template <typename gradient_type>
 SERAC_HOST_DEVICE auto abs(dual<gradient_type> x)
 {
   return (x.value >= 0) ? x : -x;
 }
 
+/**
+ * @brief Implementation of max for dual numbers
+ * @note This is not differentiable at a == b
```
Comment (white238) [inline src/serac/numerics/functional/dual.hpp]: Would these be better as `\overload brief description`?

File: src/serac/numerics/functional/dual.hpp [hunk 4]
```diff
@@ -199,13 +199,80 @@ SERAC_HOST_DEVICE constexpr auto& operator-=(dual<gradient_type>& a, double b)
...
+template <typename gradient_type>
+SERAC_HOST_DEVICE auto max(dual<gradient_type> a, double b)
+{
+  auto b_dual = dual{b, 0.0};
+  return (a > b_dual) ? a : b_dual;
+}
+
+/**
+ * @brief Implementation of max for dual numbers
+ * @note This is not differentiable at a == b
```
Comment (samuelpmishLLNL) [inline src/serac/numerics/functional/dual.hpp]: this note is good, maybe we could also say what it does in the case of equality (i.e. just returns the derivative of the second argument)

Another option might be to average the derivatives of the two operands when equal (this is probably overthinking things)

**Review by samuelpmishLLNL — APPROVED** [review]

**Pull request merged**